### ISSUE TITLE
Issues with `eachAsync` in JS driver.

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#5771]
While using `eachAsync`, I ran into two places where the docs disagreed with the behavior:

1. The docs say "`eachAsync` always returns a promise that will be resolved once all rows are returned.", but in fact `eachAsync` only returns a promise if the `final` callback is unspecified.

2. The docs say "If you do not use `rowFinished`, the `rowProcess` callback can end iteration early by returning any value other than a Promise; that value will be wrapped in an error object and passed to `final` if it is provided.", but instead `final` seems to always receive the same error: "Row handler for eachAsync may only return a Promise or undefined."

(Note that I'm not really a JS user, so it's possible I'm misunderstanding something with regard to one or both of these.)

@deontologician -- are these driver bugs or documentation bugs?

**Comment by deontologician** [comment]
ping @marshall007

**Comment by larkost** [comment]
These are both deliberate, and were the subject of much debate in #5413 (there are others, but that mostly covers it).

**Comment by marshall007** [comment]
As @larkost said, (1) is intentional based on what we discussed in #5413 though I think it's still debatable whether or not that's desirable behavior. For example, given that you can return a Promise from `final`, where are you supposed to handle that promise? You'd expect to be able to throw a `.catch(...)` on the outer promise, but we don't return the promise when `final` is specified.

(2) is just a documentation issue. We decided to only allow returning `undefined` or a `Promise` and providing a good error message so users couldn't accidentally abort iteration early by returning some "special value" (previously `false`).

As a general note, most examples still push users towards the old callback style for everything. The docs also don't really illustrate an example of the Promise-based API as it was intended to be used:

```js
cursor.eachAsync(row => {
  return processRowAsync(row) // returns a Promise
})
.then(() => {
  console.log('done processing')
})
.catch(err => {
  console.error('Error:', err.message)
});
```

**Comment by danielmewes** [comment]
Opened https://github.com/rethinkdb/docs/issues/1134 to correct 2) in the docs.

The first example in http://rethinkdb.com/api/javascript/each_async/ shows the all-promise pattern with a `.then` and `.catch`. We should probably replace other uses of callback-style iteration throughout the documentation over time.

**Comment by cacois** [comment]
Not sure if I'm seeing the same issue, but even when using the desired promise style: 

    cursor.eachAsync(row => {
        return processRowAsync(row) // returns a Promise
    })
    .then(() => {
        console.log('done processing')
    });

where processRowAsync(row) returns the result of a r.table(...).get(...) call (which is a Promise), I'm still seeing the "Row handler for eachAsync may only return a Promise or undefined" exception. 

Am I missing something here?

**Comment by larkost** [comment]
@cacois: you are using ES6 notation there, what happens if you just use the older `function (row) {}` notation? My initial thought is that it should not matter, but I am mostly trapped in working with the older notation, so there could be subtleties that I are not picking up there.

Note that I am biting my tongue hard here, but I have been in favor of handling promises in almost all of these cases.

**Comment by marshall007** [comment]
@cacois also make sure you're calling `.run()` on the inner query you're returning. `r.table(...).get(...)` in itself is not a Promise.

**Comment by cacois** [comment]
@marshall007 My fault for not being clear. I'm actually calling: 

return r.table(...).get(id).update({...}).run(this.conn);

Which, from testing I can verify does return a Promise. Still get the same exception.

**Comment by marshall007** [comment]
@cacois ah, yea the Promise you return has to resolve to `undefined` as well.

**Comment by cacois** [comment]
@marshall007 Not sure I follow - what does the Promise returned by run() resolve to?

**Comment by marshall007** [comment]
@cacois I think you need to do something like this:

```js
cursor.eachAsync(row => {
    return processRowAsync(row)
      .then(() => undefined) // prevent "may only return a Promise or undefined"
});
```

Currently, `eachAsync` checks the result of the promise you're returning to make sure you're not accidentally throwing away values you might care about from inner promises. I think it's mostly just leading to confusion though and I think we should consider changing it.

**Comment by cacois** [comment]
Ah, that was it. Thanks!

(and agreed, the eachAsync behavior is a bit confusing)

**Comment by danielmewes** [comment]
Yeah I'd be in favor of dropping the check on what a returned promise resolves to. I think it's not very helpful and makes it harder to work with `eachAsync`.

(pinging also @larkost @nighelles )

**Comment by mlucy** [comment]
> Yeah I'd be in favor of dropping the check on what a returned promise resolves to. I think it's not very helpful and makes it harder to work with eachAsync.

It's mildly helpful because it prevents you from thinking that the promise you return inside `eachAsync` is going somewhere, and encourages you to chain on a `.then` to check for errors if you're running ReQL queries inside the `eachAsync`, but I agree that it probably causes more pain than it prevents, since sometimes it won't fail until you hit a particular branch at runtime.
